dipimage/findlocmax.m

```diff
@@ -0,0 +1,176 @@
+% September 2004, nD subpixel finding with own size
+% October 2004, added support for negative values, where center of mass does not work
+% April, 2005, return max value at maximum position
+% April, 2005, added several subpixel methods
+% July 2006, remove local background for CM, still the best way is to go with an object
+%            mask + Gravity, the square box is not optimal, especially for skew intensity
+% 27 September 2010: Rewritten to use SUBSREF instead of EVAL. (CL)
+% 23 December 2010: Changed finding of new coordinate, before gave sometimes strange results (BR)
+% 21 July 2011, Catch case when local max is on the boundary (BR)
+
+function [out,out2] = findlocmax(in,cood,lsize,subp,subp_lsize,submeth)
+
+if nargin < 2
+    cood = [20 160];
+end
+if nargin < 3
+    lsize = [15 15];
+end
+if nargin < 4
+    subp = 0;
+end
+if nargin < 5
+    subp_lsize = [0 0];
+end
+if nargin < 6
+    submeth = 'parabolic';
+end
+
+%give possibility to supply not int cood, but round it here
+cood = round(cood);
+in = squeeze(in);%just make sure
+if any(cood > size(in))
+    error('Coordinates not within the image.');
+end
+
+
+N = ndims(in);
+s = floor(lsize./2);
+%2s+1 neighborhood
+co = zeros(N,2);%corner array
+ss = substruct('()',cell(1,N));
+ts = substruct('()',cell(1,N));
+
+%fprintf('cood: %d %d, s: %d %d\n',cood,s);
+done =0;
+while ~done
+    for ii=1:N
+        co(ii,1) = max(min(cood(ii)-s(ii),size(in,ii)-1),0);
+        co(ii,2) = max(min(cood(ii)+s(ii),size(in,ii)-1),0);
+        ss.subs{ii} = co(ii,1):co(ii,2);
+        ts.subs{ii} = cood(ii);
+    end
+    roi = subsref(in,ss);  % get roi
+    [v,nc] = max(roi);          % get max value in roi (and the coordinate)
+    v = double(v);
+    
+    %fprintf('max_n: %f, cood(1) %d, cood(2) %d\n',v,cood(1),cood(2))
+    tmpv = double(subsref(in,ts)); % get value of prior max
+    
+    if v == tmpv
+        out2 = v; %intensity
+        out = cood; %coordinate position
+        done = 1;
+    else
+        %nc = nc(1,:);%just in case we find more than one max
+        cood = cood - s;
+        cood(cood<0)=0;
+        cood = cood + nc;
+    end
+end
+if subp
+    %check if the initial coordinates are on the edge of the image, we
+    %cannot deal with that
+    sz=size(in);
+    if any(out >= sz-1)
+        return;
+    end
+    if ~subp_lsize
+        subp_lsize = s;
+    else
+        subp_lsize = floor(subp_lsize./2);
+    end
+    %fprintf(' out:%d %d\n',out)
+    if strcmp(submeth,'CenterOfMass')
+        mask = getcube(in,out,subp_lsize);
+        if any(in(mask)<0)
+            warning(['Negative values in the ROI of the image, center of mass estimation does not work ' ...
+                'switching to parabolic subpixel fit.']);
+            submeth = 'parabolic';
+        end
+    end
+    switch submeth
+        case 'CenterOfMass'
+            [tmp, v] = dip_isodatathreshold(in,mask,1);
+%             [tmp2, v2] = threshold(in,'isodata');

```

Comment:
Yan,
I don't like this `dip_isodatathreshold`. I have added the mask parameter to `threshold`, so we can use that function there. I can make this change.

Reply:
This was slightly more involved, but a lot nicer I think.
9f3770485a9ed5966edb9fe572e24bc70d218cee